scripts/create-camayoc-ocp-config.py

```diff
@@ -35,7 +35,11 @@ def login(args):
         cmd = ["oc", "login", url, insecure, "--token", args.auth_token]
     else:
         cmd = ["oc", "login", url, insecure, "-u", args.username, "-p", args.password]
-    result = subprocess.run(cmd, capture_output=True, check=False, text=True)
+    try:
+        result = subprocess.run(cmd, capture_output=True, check=False, text=True)
+    except TypeError:
```

Comment:
Now you got me curious - on what conditions can subprocess raise `TypeError`? Usually I see `TimeoutExpired` or `CalledProcessError` being handled.

I looked briefly at [subprocess source code](https://github.com/python/cpython/blob/main/Lib/subprocess.py) and I see few place where  `TypeError` is raised, but I don't think we should hit any of them here. Maybe it is raised implicitly somewhere.

Reply:
It happens when the password or the token is empty (None).

```
❯ ./scripts/create-camayoc-ocp-config.py  -a localhost      
Traceback (most recent call last):
  File "/home/rmoura/Sources/camayoc/./scripts/create-camayoc-ocp-config.py", line 168, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/rmoura/Sources/camayoc/./scripts/create-camayoc-ocp-config.py", line 159, in main
    if not login(args):
           ^^^^^^^^^^^
  File "/home/rmoura/Sources/camayoc/./scripts/create-camayoc-ocp-config.py", line 38, in login
    result = subprocess.run(cmd, capture_output=True, check=False, text=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.11/subprocess.py", line 1883, in _execute_child
    self.pid = _fork_exec(
               ^^^^^^^^^^^
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

```
❯ ./scripts/create-camayoc-ocp-config.py  -a localhost -t ''
Traceback (most recent call last):
  File "/home/rmoura/Sources/camayoc/./scripts/create-camayoc-ocp-config.py", line 168, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/rmoura/Sources/camayoc/./scripts/create-camayoc-ocp-config.py", line 159, in main
    if not login(args):
           ^^^^^^^^^^^
  File "/home/rmoura/Sources/camayoc/./scripts/create-camayoc-ocp-config.py", line 38, in login
    result = subprocess.run(cmd, capture_output=True, check=False, text=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.11/subprocess.py", line 1883, in _execute_child
    self.pid = _fork_exec(
               ^^^^^^^^^^^
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

Reply:
Diving into the code:

* https://github.com/python/cpython/blob/main/Lib/subprocess.py#L104
* https://github.com/python/cpython/blob/main/Modules/_posixsubprocess.c#L1007

Reply:
I see. We only check token here, and if it's falsy, we take username and pass path - but these are not guaranteed to be present, either.

I think another approach would be to prevent it with something like:
```
    if args.auth_token:
        cmd = ["oc", "login", url, insecure, "--token", args.auth_token]
    elif args.username and args.password:
        cmd = ["oc", "login", url, insecure, "-u", args.username, "-p", args.password]
    else:
        print("you need to pass token or (username and password)!"
        return False
```

Thanks, now I understand the issue.